actors/builtin/miner/deadline_state.go

```diff
@@ -880,14 +926,29 @@ func (p *PoStResult) PenaltyPower() PowerPair {
 // changes to power (newly faulty power, power that should have been proven
 // recovered but wasn't, and newly recovered power).
 //
-// NOTE: This function does not actually _verify_ any proofs. The returned
-// Sectors and IgnoredSectors must subsequently be validated against the PoSt
-// submitted by the miner.
+// NOTE: This function does not actually _verify_ any proofs.
 func (dl *Deadline) RecordProvenSectors(
 	store adt.Store, sectors Sectors,
 	ssize abi.SectorSize, quant QuantSpec, faultExpiration abi.ChainEpoch,
 	postPartitions []PoStPartition,
+	proofs []proof.PoStProof,
 ) (*PoStResult, error) {
+
+	partitionIndexes := bitfield.New()
+	for _, partition := range postPartitions {
+		partitionIndexes.Set(partition.Index)
+	}
+
+	// First check to see if we're proving any already proven partitions.
+	// This is faster than checking one by one.
+	if alreadyProven, err := bitfield.IntersectBitField(dl.PartitionsPoSted, partitionIndexes); err != nil {
+		return nil, xerrors.Errorf("failed to check proven partitions: %w", err)
+	} else if empty, err := alreadyProven.IsEmpty(); err != nil {
+		return nil, xerrors.Errorf("failed to check proven intersection is empty: %w", err)
+	} else if !empty {
+		return nil, xc.ErrIllegalArgument.Wrapf("partition already proven: %v", alreadyProven)
+	}
+
```

Comment:
By moving this check outside of the loop over `postPartitions`, it becomes possible for `postPartitions` to contain duplicate indices (though `post.skipped` may not be duplicate!).

I couldn't find any route to this actually messing up accounting, but there are so many options to think about. At a minimum, the tests needed to effectively cover this new behavior would be quite extensive.

I would recommend moving this back in the loop below, or otherwise explicitly disallowing the miner submitting duplicate indices.

Reply:
Hm. You're right. I think we at least need to make sure there aren't duplicates. Those aren't allowed anyways.